streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -16,10 +16,298 @@
+    }
+
+    public boolean hasNoNonGlobalTopology() {
+        return evaluateConditionIsTrueForAnyBuilders(InternalTopologyBuilder::hasNoNonGlobalTopology);
+    }
+
+    public boolean hasPersistentStores() {
+        // If the app is using named topologies, there may not be any persistent state when it first starts up
+        // but a new NamedTopology may introduce it later, so we must return true
+        if (hasNamedTopologies()) {
+            return true;
+        }
+        return evaluateConditionIsTrueForAnyBuilders(InternalTopologyBuilder::hasPersistentStores);
+    }
+
+    public boolean hasStore(final String name) {
+        return evaluateConditionIsTrueForAnyBuilders(b -> b.hasStore(name));
+    }
+
+    public boolean hasOffsetResetOverrides() {
+        return evaluateConditionIsTrueForAnyBuilders(InternalTopologyBuilder::hasOffsetResetOverrides);
+    }
+
+    public OffsetResetStrategy offsetResetStrategy(final String topic) {
+        for (final InternalTopologyBuilder builder : builders.values()) {
+            final OffsetResetStrategy resetStrategy = builder.offsetResetStrategy(topic);
+            if (resetStrategy != null) {
+                return resetStrategy;
+            }
+        }
+        return null;
+    }
+
+    Collection<String> sourceTopicCollection() {
+        final List<String> sourceTopics = new ArrayList<>();
+        applyToEachBuilder(b -> sourceTopics.addAll(b.sourceTopicCollection()));
+        return sourceTopics;
+    }
+
+    Pattern sourceTopicPattern() {
+        final StringBuilder patternBuilder = new StringBuilder();
+
+        applyToEachBuilder(b -> {
+            final String patternString = b.sourceTopicsPatternString();
+            if (patternString.length() > 0) {
+                patternBuilder.append(patternString).append("|");
+            }
+        });
+
+        if (patternBuilder.length() > 0) {
+            patternBuilder.setLength(patternBuilder.length() - 1);
+            return Pattern.compile(patternBuilder.toString());
+        } else {
+            return EMPTY_ZERO_LENGTH_PATTERN;
+        }
+    }
+
+    public boolean usesPatternSubscription() {
+        return evaluateConditionIsTrueForAnyBuilders(InternalTopologyBuilder::usesPatternSubscription);
+    }
+
+    // Can be empty if app is started up with no Named Topologies, in order to add them on later
+    public boolean isEmpty() {
+        return builders.isEmpty();
+    }
+
+    public String topologyDescription() {
+        if (isEmpty()) {
+            return "";
+        }
+        final StringBuilder sb = new StringBuilder();
+
+        applyToEachBuilder(b -> {
+            sb.append(b.describe().toString());
+        });
+
+        return sb.toString();
+    }
+
+    public final void rewriteAndBuildTopology(final StreamsConfig config) {
+        // As we go, check each topology for overlap in the set of input topics/patterns
+        final Set<String> allInputTopics = new HashSet<>();
```

Comment:
Well we should only need to call this once -- for now. As it happens I did actually move this to a class field in Pt. 3 since by then we'll be adding new topologies that need to be verified